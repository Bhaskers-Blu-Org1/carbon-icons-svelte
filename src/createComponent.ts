import {
  formatAttributes,
  getAttributes,
  toString
} from '@carbon/icon-helpers';

interface IAttributes {
  focusable: 'false';
  preserveAspectRatio: 'xMidYMid meet';
  style: 'will-change: transform;';
  width: undefined;
  height: undefined;
  viewBox: '0 0 undefined undefined';
  'aria-hidden': true;
}

const attributes: IAttributes = getAttributes();

type IconSize = 16 | 20 | 24 | 32;

interface IPath {
  elem: 'path';
  attrs: { d: string };
}

interface ICircle {
  elem: 'circle';
  attrs: { cx: string; cy: string; r: string };
}

interface IRect {
  elem: 'rect';
  attrs: { width: string; height: string; x: string; y: string; rx: string };
}

export type CarbonIconContent = ReadonlyArray<IPath | ICircle | IRect>;

export interface ICarbonIcon {
  elem: 'svg';
  attrs: {
    xmlns: 'http://www.w3.org/2000/svg';
    viewBox: string;
    width: IconSize;
    height: IconSize;
  };
  content: CarbonIconContent;
  name: string;
  size?: IconSize;
}

function createComponent(icon: ICarbonIcon) {
  const { name, size, attrs, content } = icon;

  const markup = `
    <script>
      let className = undefined;
      export { className as class };
      export let tabindex = undefined;
      export let focusable = ${attributes.focusable};
      export let title = undefined;
      export let style = '${attributes.style}';

      const preserveAspectRatio = '${attributes.preserveAspectRatio}';
      const ariaLabel = $$props['aria-label'];
      const ariaLabelledBy = $$props['aria-labelledby'];
      const labelled = ariaLabel || ariaLabelledBy || title;
      const attributes = {
        'aria-label': ariaLabel,
        'aria-labelledby': ariaLabelledBy,
        'aria-hidden': labelled ? undefined : true,
        role: labelled ? 'img' : undefined,
        focusable: tabindex ? true : focusable,
        tabindex
      };
    </script>
    <svg
      on:click
      on:mouseover
      on:mouseenter
      on:mouseleave
      ${formatAttributes(attrs)}
      class={className}
      {preserveAspectRatio}
      {style}
      {...attributes}>
      ${content.map(element => toString(element)).join('')}
      <slot>
        {#if title}
          <title>{title}</title>
        {/if}
      </slot>
    </svg>
  `;

  return { name, size, markup };
}

export { createComponent };
