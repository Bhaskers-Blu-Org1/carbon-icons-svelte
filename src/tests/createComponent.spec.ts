import * as icons from '@carbon/icons';
import { createComponent } from '../createComponent';

test('icon matches the snapshot', () => {
  const icon = createComponent(icons.ArrowLeft16);
  expect(icon).toMatchSnapshot();
});
