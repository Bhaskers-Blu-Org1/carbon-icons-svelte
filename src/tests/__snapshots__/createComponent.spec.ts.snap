// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon matches the snapshot 1`] = `
Object {
  "markup": "
    <script>
      let className = undefined;
      export { className as class };
      export let tabindex = undefined;
      export let focusable = false;
      export let title = undefined;
      export let style = 'will-change: transform;';

      const preserveAspectRatio = 'xMidYMid meet';
      const ariaLabel = $$props['aria-label'];
      const ariaLabelledBy = $$props['aria-labelledby'];
      const labelled = ariaLabel || ariaLabelledBy || title;
      const attributes = {
        'aria-label': ariaLabel,
        'aria-labelledby': ariaLabelledBy,
        'aria-hidden': labelled ? undefined : true,
        role: labelled ? 'img' : undefined,
        focusable: tabindex ? true : focusable,
        tabindex
      };
    </script>
    <svg
      on:click
      on:mouseover
      on:mouseenter
      on:mouseleave
      xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\" width=\\"16\\" height=\\"16\\"
      class={className}
      {preserveAspectRatio}
      {style}
      {...attributes}>
      <path d=\\"M6.7 12.3L2.9 8.5H15v-1H2.9l3.8-3.8L6 3 1 8l5 5z\\"></path>
      <slot>
        {#if title}
          <title>{title}</title>
        {/if}
      </slot>
    </svg>
  ",
  "name": "arrow--left",
  "size": 16,
}
`;
